Widget Visibility: add ability to match ALL conditions.

zinigor opened the pull request:
This fixes issue #852. (Replay of #3289 by @atrus1701 )

    Add the option to match all conditions specified for the widget.

This is related to Pull Request #3210. This original Pull Request contained several fixes that are being broken out into several smaller requests.

zinigor commented:
Master issue: #5781

zinigor commented:
I have added a couple of fixes:
* The `match_all` flag would not work properly if any of the conditions were met, now the logic checks all conditions.
* The checkbox is now under the conditions and has some padding.

georgestephanis commented on modules/widget-visibility/widget-conditions.php:
```diff
@@ -560,7 +574,7 @@ public static function filter_widget( $instance ) {
 							break;
 							case 'posts':
-								$condition_result = $wp_query->is_posts_page;
+								$condition_result = is_single();
```
This is a definite change -- `is_posts_page` returns true for the posts archive view when there's a page on the site's home.  `is_single()` returns true on a page's permalink, non-archive view.

https://github.com/WordPress/WordPress/blob/9d7ea04936125041b224b227f48ad4194066b314/wp-includes/class-wp-query.php#L1008

georgestephanis commented on modules/widget-visibility/widget-conditions.php:
```diff
@@ -349,7 +352,11 @@ public static function widget_conditions_admin( $widget, $return, $instance ) {
-								<span class="condition-conjunction"><?php echo esc_html_x( 'or', 'Shown between widget visibility conditions.', 'jetpack' ); ?></span>
+								<span class="condition-conjunction">
+									<?php
+										echo esc_html_x( 'or', 'Shown between widget visibility conditions.', 'jetpack' );
```
This probably shouldn't be broken out onto it's own line from the php opening and closing tags.

georgestephanis commented on modules/widget-visibility/widget-conditions.php:
```diff
@@ -301,6 +301,9 @@ public static function widget_conditions_admin( $widget, $return, $instance ) {
 		if ( empty( $conditions['rules'] ) )
 			$conditions['rules'][] = array( 'major' => '', 'minor' => '', 'has_children' => '' );
 
+		if ( ! isset( $conditions['match_all'] ) )
```
Conditionals need braces.

georgestephanis replied:
Also, may be better to use `empty()` here.

georgestephanis commented on modules/widget-visibility/widget-conditions.php:
```diff
@@ -577,7 +591,7 @@ public static function filter_widget( $instance ) {
 								} elseif ( $rule['minor'] == get_option( 'page_for_posts' ) ) {
 									// If $rule['minor'] is a page ID which is also the posts page
-									$condition_result = $wp_query->is_posts_page;
+									$condition_result = is_single();
```
Same concern as above.  Breaking existing behavior by testing for something different.

georgestephanis commented on modules/widget-visibility/widget-conditions.php:
```diff
@@ -679,12 +693,39 @@ public static function filter_widget( $instance ) {
+
+				// In case the match_all flag was set we quit on first failed condition
+				break;
+			} elseif(
```
Needs a space between `elseif` and `(`

georgestephanis commented on modules/widget-visibility/widget-conditions.php:
```diff
@@ -679,12 +693,39 @@ public static function filter_widget( $instance ) {
+				break;
+			} elseif(
+				(
+					! isset( $instance['conditions']['match_all'] )
```
Again, may be more readable as `empty()` instead of negating.

georgestephanis commented on modules/widget-visibility/widget-conditions.php:
```diff
@@ -679,12 +693,39 @@ public static function filter_widget( $instance ) {
+				'show' == $instance['conditions']['action']
+				&& ! $condition_result
+			)
+			|| (
```
I personally find broken out conditionals like this to be harder to read, but if we do, can we please do it as `) || (` instead?

zinigor commented:
Thanks for the review George! I have fixed the problems you have found. As with the checkbox - I think we need a design review here. This is what the condition table currently looks like:
![conditions](https://cloud.githubusercontent.com/assets/374293/23227241/b97090ee-f951-11e6-8fb8-c7aaba40276c.png)

zinigor commented:
For reference that's what it looks like in `master`. I see that the styles and images for some things are missing, I'll fix that. But the main thing is that we need to figure out where to put the checkbox best.
![master](https://cloud.githubusercontent.com/assets/374293/23227501/95d17ea4-f952-11e6-8795-31f5ae72717a.png)

zinigor commented:
I have rebased this PR and added the following changes:
* JSHint no longer has any warnings about the JS file;
* The delimiter between the conditions now displays "or"/"and" based on the status of the "Match all" checkbox:

![deselected](https://cloud.githubusercontent.com/assets/374293/23359054/ef78835e-fcfd-11e6-9a6c-2869a97c0623.png)

![selected](https://cloud.githubusercontent.com/assets/374293/23359058/f2efb6f6-fcfd-11e6-959e-e45dda97430e.png)

Since @MichaelArestad has given this one a :+1: , I'm removing the Desing Needed label, and since @georgestephanis has given this a :+1: , I'm putting the PR into Ready to Merge mode.

dereksmart commented on modules/widget-visibility/widget-conditions.php:
```diff
@@ -596,12 +638,38 @@ public static function filter_widget( $instance ) {
+			} elseif (
+				(
+					empty( $instance['conditions']['match_all'] )
+					|| $instance['conditions']['match_all'] !== '1'
```
should be yoda style

dereksmart commented on modules/widget-visibility/widget-conditions.php:
```diff
@@ -596,12 +638,38 @@ public static function filter_widget( $instance ) {
-			if ( $condition_result )
+			if (
+				isset( $instance['conditions']['match_all'] )
+				&& $instance['conditions']['match_all'] == '1'
```
yoda style please
